### DISCUSSION TITLE
Database migrations and warning

### DISCUSSION THREAD

**Discussion by Epy** [mastodon/mastodon#18412] [Q&A] [answered]
Hello
I was about to update my Mastodon instance to the latest 3.5 (I'm in 3.4.6) and saw a warning message on admin dashboard:

![Capture d’écran du 2022-05-14 10-37-41](https://user-images.githubusercontent.com/450856/168420585-f1463e00-1d77-47b5-9f62-0c7363805438.png)
(Saying there is database migrations waiting and I should execute them for the application to execute properly)

As I already tried in the past doing such "removing unused indexes" as pgHero says, I search for infos tried the first command told in pgHero/Space and was about to ask for some help with the next commands when I found this:
https://github.com/mastodon/mastodon/issues/17768

Gargron says Mastodon/PgHero shoud hide the migration recommandations because it has caused problems to Mastodon admins.
Is this yellow message useful then ? (I'm on production mode)

Am I safe to update to 3.5.x or should I check some things before ?

Thanks

**Comment by Gargron** [comment]
This isn't pghero. This is *our* dashboard.

**Comment by Epy** [comment]
I was talking about both: in previous Mastodon versions I saw the advice to remove unused indexes only on pgHero dashboard, and tried to follow instructions. In this version (maybe a few previous ? I don't go often on the Mastodon admin dashboard) I found this yellow notification.
I am curious why there is this yellow one if such instructions shouldn't be followed on production databases as you seem to explain.
Thus, should I ignore it or may it indicate an incoming issue ?

**Reply by nightpool** [reply]
@Epy the two warnings are complicated unrelated. PgHero warnings about unused indexes are not important. Mastodon warnings about pending migrations are very important, and they indicate that you may have messed up the upgrade instructions.

However, technically, it looks like the warning *can* appear in the middle of an upgrade even if you're doing everything right. in that case it means that you can't yet restart your web server until you finish running migrations (so you're in a state where if you restarted your web server, things would break)

**Comment by Epy** [comment, marked as answer]
Oh thanks for the info, I thought they were related (because of the first try to execute PgHero instructions a while ago)

I did restart the server a few times recently without any problems, so I may not be in the state you think ?

Can someone help me finding which migration is waiting and executing it ? I'm pretty sure it's not an upgrade database migration, I follow instructions seriously each time since the installation (in 3.2 or 3.3 I don't remember).
If I shoud ask for help somewhere else or create an issue please tell me :)

Thanks

**Reply by ClearlyClaire** [reply]
What you have to run is `RAILS_ENV=production bundle exec rails db:migrate`

**Reply by Epy** [reply]
Thank you

This is the answer I get:

```
$ RAILS_ENV=production RUBYOPT=-W0 bundle exec rails db:migrate
rails aborted!
ActiveRecord::DuplicateMigrationNameError: 

Multiple migrations have the name RemoveUnusedIndexes.

/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/activerecord-6.1.3.2/lib/active_record/migration.rb:1353:in `validate'
/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/activerecord-6.1.3.2/lib/active_record/migration.rb:1226:in `initialize'
/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/activerecord-6.1.3.2/lib/active_record/migration.rb:1086:in `new'
/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/activerecord-6.1.3.2/lib/active_record/migration.rb:1086:in `up'
/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/activerecord-6.1.3.2/lib/active_record/migration.rb:1061:in `migrate'
/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/activerecord-6.1.3.2/lib/active_record/tasks/database_tasks.rb:237:in `migrate'
/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/strong_migrations-0.7.6/lib/strong_migrations/database_tasks.rb:4:in `migrate'
/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/activerecord-6.1.3.2/lib/active_record/railties/databases.rake:92:in `block (3 levels) in <top (required)>'
[SNIP repeated errors]
/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/activesupport-6.1.3.2/lib/active_support/dependencies.rb:332:in `block in require'
/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/activesupport-6.1.3.2/lib/active_support/dependencies.rb:299:in `load_dependency'
/home/mastodon/live/vendor/bundle/ruby/2.7.0/gems/activesupport-6.1.3.2/lib/active_support/dependencies.rb:332:in `require'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

I also typed `git status` in the ~/live directory and got three files which are not from Mastodon repository:
```$ git status
HEAD détachée sur v3.4.6
Fichiers non suivis: (not followed)

	db/migrate/20200725144904_remove_unneeded_indexes.rb.bak
	db/migrate/20220112134124_remove_unused_indexes.rb
	db/migrate/20220514091353_remove_unused_indexes2.rb


```

How to remove these useless migrations (now I know I shouldn't execute them)
Should I just delete the files or is there something else to do before ?

**Reply by ClearlyClaire** [reply]
Yes, delete the files.

**Reply by Epy** [reply]
The warning message was still there, but it disappeared after a new run of `rails db:migrate` !  (FixCanonicalEmailBlocksForeignKey was executed)

My tries and mistake of removing unused indexes had blocked this migration in fact.

Thank you for your help !

I'll now update to 3.5.2

Edit: I can't mark your answer as an answer, I'll mark mine..
